### ISSUE TITLE
The network fallback should be rejected when the request body is locked or disturbed

### ISSUE THREAD

**Issue by yutakahirano** [w3c/ServiceWorker#1560]
At the 24th step In https://w3c.github.io/ServiceWorker/#on-fetch-request-algorithm, we should reject the request if the request body is non-null, and it is locked or disturbed, because otherwise we cannot pass the streaming body to the network module.

**Comment by yutakahirano** [comment]
@annevk @yoichio

**Comment by annevk** [comment]
It might be cleaner to do this in Fetch when the service worker returns null? Or is there a case I'm missing?

**Comment by yutakahirano** [comment]
I'm fine with that too.

**Comment by annevk** [comment]
I think Fetch should make a copy of request before handing it to the service worker as well. I think as written it currently breaks down the moment you do `req.blob()` for non-stream requests in a service worker.

**Comment by yutakahirano** [comment]
>  I think as written it currently breaks down the moment you do req.blob() for non-stream requests in a service worker.

Sorry I don't understand this sentence. Could you rephrase it?

**Comment by annevk** [comment]
If the service worker invokes `req.blob()` in Handle Fetch and then Handle Fetch ends up returning null to Fetch, Fetch holds a request whose body is disturbed. This is problematic as Fetch still needs that body. Therefore, similar to making an HTTP request (where you can have a similar situation if you encounter a redirect) it seems that Fetch will need to copy the request.

**Comment by yutakahirano** [comment]
I think it's better to fail the request when the request body got locked or disturbed.

**Comment by annevk** [comment]
Including in the non-streaming case? What do user agents do today?

**Comment by yutakahirano** [comment]
With non-streaming body Chromium copies the request body. With streaming body it crashes.

If there are no compatibility concerns I think failing the request always is a more consistent (and hence better) behavior. I'm trying to measure the impact of that change: https://chromium-review.googlesource.com/c/chromium/src/+/2631550

**Comment by annevk** [comment]
Well note that we already have this difference for HTTP-network fetches as I tried to point out. We copy for non-streaming and error for streaming when we end up having to reuse a body. It makes sense somewhat to align SW with that and I doubt it would be compatible to change that now.

**Comment by yutakahirano** [comment]
Thanks, I'll try to make a PR.
